### PR TITLE
solution added

### PR DESCRIPTION
**Pull request by IrynaDim**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -8,5 +8,69 @@
...
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
 
 </project>
```
Comment (DmGrt) [inline pom.xml]: Add empty line

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                Storage.products.remove(i);
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean delete(Product product) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i) == product) {
```
Comment (DmGrt) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: here better to use equals for Products

**Review by DmGrt — COMMENTED** [review]
Great job! Just correct a little bit

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java (hunk 2 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: точно)) то випадково щось затупила) чогось вирішила що то цифри)))))

File: pom.xml [hunk 3]
```diff
@@ -8,5 +8,69 @@
...
 
-</project>
\ No newline at end of file
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (KseniiaMakarova) [inline pom.xml]: let's remove it - currently we don't use it

File: src/main/java/com/internet/shop/dao/ProductDao.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    Optional<Product> getById(Long productId);
+
+    Product update(Product product);
+
+    boolean deleteById(Long productId);
+
+    boolean delete(Product product);
+
+    List<Product> getAllProducts();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDao.java]:
```suggestion
    boolean delete(Long productId);
    
    List<Product> getAll();
```

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public Optional<Product> getById(Long productId) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId() == productId) {
+                return Optional.ofNullable(Storage.products.get(i));
+            }
+        }
+        return Optional.empty();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: let's use findFirst()

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package com.internet.shop.dao;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]:
```suggestion
package com.internet.shop.dao.impl;
```

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        return Optional.empty();
+    }
+
+    @Override
+    public Product update(Product product) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId() == product.getId()) {
+                Storage.products.set(i, product);
+            }
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: the logic is ok, but let's use Stream API

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public boolean deleteById(Long productId) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId() == productId) {
+                Storage.products.remove(i);
+            }
+        }
+        return true;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: we can use `removeIf()` on storage

File: src/main/java/com/internet/shop/db/Storage.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    public static final List<Order> orders = new ArrayList<>();
+    public static final List<ShoppingCart> shoppingCarts = new ArrayList<>();
+    public static final List<User> users = new ArrayList<>();
+    private static Long productId = 0L;
+    private static Long orderId = 0L;
+    private static Long shoppingCartId = 0L;
+    private static Long userId = 0L;
+
+    public static void addProduct(Product product) {
+        product.setId(productId++);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/db/Storage.java]:
```suggestion
        product.setId(++productId);
```
ids should start with 1

File: src/main/java/com/internet/shop/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public class Order {
+    private Long orderId;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(Long orderId, List<Product> products, Long userId) {
+        this.orderId = orderId;
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: id is set in storage
and we won't receive product list from outside

File: src/main/java/com/internet/shop/model/Product.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package com.internet.shop.model;
+
+import java.math.BigDecimal;
+
+public class Product {
+    private Long id;
+    private String name;
+    private BigDecimal price;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Product.java]: let's make this field double for now

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(Long id, List<Product> products, Long userId) {
+        this.id = id;
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: id is set in storage
and we won't receive product list from outside

File: src/main/java/com/internet/shop/model/User.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
+package com.internet.shop.model;
+
+public class User {
+    private Long id;
+    private String name;
+    private String login;
+    private String password;
+
+    public User(Long id, String name, String login, String password) {
+        this.id = id;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/User.java]: id is set in storage

File: src/main/java/com/internet/shop/service/ProductService.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    Product get(Long id);
+
+    Product update(Product product);
+
+    boolean deleteById(Long productId);
+
+    boolean delete(Product product);
+
+    List<Product> getAllProducts();
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/ProductService.java]:
```suggestion
    boolean delete(Long productId);

    List<Product> getAll();
```

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    private ProductDao productDao;
+
+    @Override
+    public Product create(Product product) {
+        return productDao.create(product);
+    }
+
+    @Override
+    public Product get(Long id) {
+        return productDao.getById(id).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]: let's use `get()` for now

File: src/main/java/com/internet/shop/Application.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+import com.internet.shop.service.ProductService;
+import java.math.BigDecimal;
+
+public class Application {
+    private static Injector injector = Injector.getInstance("com.internet.shop");
+
+    public static void main(String[] args) {
+        ProductService productService = (ProductService) injector.getInstance(ProductService.class);
+        Product product = new Product("Teapot", new BigDecimal(320.0));
+        Product product1 = new Product("Cup", new BigDecimal(80.0));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java]: make your product names informative

File: src/main/java/com/internet/shop/Application.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+        productService.create(product);
+        productService.create(product1);
+        productService.create(new Product("Spoon", new BigDecimal(25.0)));
+        productService.create(new Product("Fork", new BigDecimal(35.0)));
+        productService.create(new Product("Knife", new BigDecimal(50.0)));
+        productService.delete(product);
+
+        productService.get(4L);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java]:
```suggestion
        productService.get(someProduct,getId());
```
let's avoid hardcode

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/dao/ProductDao.java (hunk 4 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/dao/ProductDao.java]: +

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java (hunk 6 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: +

File: src/main/java/com/internet/shop/service/ProductService.java (hunk 14 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/service/ProductService.java]: done

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java (hunk 15 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]: done. if i understant correct))

File: src/main/java/com/internet/shop/Application.java (hunk 16 above)
Comment (IrynaDim) [reply src/main/java/com/internet/shop/Application.java]: done

File: pom.xml [hunk 18]
```diff
@@ -8,5 +8,63 @@
...
+        <maven.checkstyle.plugin.version>3.1.1</maven.checkstyle.plugin.version>
+        <maven.checkstyle.plugin.configLocation>
+            https://raw.githubusercontent.com/mate-academy/style-guides/master/java/checkstyle.xml
+        </maven.checkstyle.plugin.configLocation>
+    </properties>
 
-</project>
\ No newline at end of file
+    <dependencies>
+    </dependencies>
```
Comment (KseniiaMakarova) [inline pom.xml]: we can remove this

File: src/main/java/com/internet/shop/dao/ProductDao.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    Product create(Product product);
+
+    Optional<Product> get(Long productId);
+
+    Product update(Product product);
+
+    boolean delete(Long productId);
+
+    List<Product> getAllProducts();
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDao.java]:
```suggestion
```

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    public Product create(Product product) {
+        Storage.addProduct(product);
+        return product;
+    }
+
+    @Override
+    public Optional<Product> get(Long productId) {
+        return Optional.ofNullable(Storage.products.stream()
+                .filter(product -> product.getId() == productId)
+                .findFirst().orElseThrow());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]:
```suggestion
        return Storage.products.stream()
                .filter(product -> product.getId() == productId)
                .findFirst();
```

also, do we use `==` to compare object types?

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+    public Product update(Product product) {
+        Storage.products.stream()
+                .filter(x -> x.getId() == product.getId())
+                .map(x -> product);
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long productId) {
+        return Storage.products.removeIf(product -> (product.getId() == productId));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: same question about `==`, check all your methods

File: src/main/java/com/internet/shop/model/Order.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.List;
+
+public class Order {
+    private Long orderId;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: we will not receive `products` from outside
remove this from method arguments and initialize as empty `ArrayList`

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: same here

File: src/main/java/com/internet/shop/service/ProductService.java [hunk 24]
```diff
@@ -0,0 +1,19 @@
+package com.internet.shop.service;
+
+import com.internet.shop.lib.Dao;
+import com.internet.shop.model.Product;
+import java.util.List;
+
+@Dao
+public interface ProductService {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/ProductService.java]: `@Dao`?

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override
+    public Product create(Product product) {
+        return productDao.create(product);
+    }
+
+    @Override
+    public Product get(Long id) {
+
+        return productDao.get(id).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]:
```suggestion
        return productDao.get(id).get();
```

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override
+    public Product update(Product product) {
+        return productDao.update(product);
+    }
+
+    @Override
+    public boolean delete(Long productId) {
+
+        return productDao.delete(productId);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]:
```suggestion
        return productDao.delete(productId);
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+        return Storage.products.stream()
+                .filter(product -> product.getId().equals(productId))
+                .findFirst();
+    }
+
+    @Override
+    public Product update(Product product) {
+        Storage.products.stream()
+                .filter(x -> x.getId().equals(product.getId()))
+                .map(x -> product);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: this doesn't cause changes in our storage
the elements are changed only in this stream

let's create IntStream of indices, than filter out only those indices on which there are elements with the same id
then for each of the remaining indices, **replace the element in storage** with our `product`

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**